content/router.xql

```diff
@@ -75,376 +36,300 @@ declare function router:route($jsonPaths as xs:string+, $lookup as function(xs:s
+            $ref (: internal spec lookup :)
+
+    return router:resolve-ref($config, $parts)
 };
 
-declare function router:process($routes as map(*)*, $lookup as function(*)) {
-    (: if there are multiple matches, prefer the one matching the longest pattern :)
-    let $route := sort($routes, (), function($route) {
-            string-length(replace($route?pattern, "\{[^\}]+\}", "?")) + $route?priority
-        }) => reverse() => head()
-    let $loginDomain := router:login-domain($route?spec)
-    let $parameters := map:merge((
-        router:map-request-parameters($route?config),
-        router:map-path-parameters($route, $route?path)
-    ))
-    let $info := $route?spec?info
-    let $request := map {
-        "parameters": $parameters,
-        "body": router:request-body($route?config),
-        "loginDomain": $loginDomain,
-        "info": $info,
-        "config": $route
+(:~
+ : Maps a request to the configured handler 
+ : Loads API definitions, matches routes
+ : Calls route handler function returned by $lookup function
+ :
+ : Route specificity rules:
+ : 1. normalize patterns: replace placeholders with "?"
+ : 2. use the matching route with the longest normalized pattern
+ : 3. If two paths have the same (normalized) length, prioritize by appearance in API files, first one wins
+ :)
+declare function router:route($api-files as xs:string+, $lookup as function(xs:string) as function(*)?, $middlewares as function(*)*) {
+    let $controller := request:get-attribute("$exist:controller")
+    let $base-collection := ``[`{repo:get-root()}`/`{$controller}`/]``
+
+    let $request-data := map { 
+        "id": util:uuid(),
+        "method": request:get-method() => lower-case(),
+        "path": request:get-attribute("$exist:path")
     }
+
     return (
-        if ($loginDomain) then (
-            login:set-user($loginDomain, (), false())
-        ) else
-            (),
-        if (router:check-login($route?config)) then
-            ()
-        else
-            error($errors:UNAUTHORIZED, "Access denied"),
-        router:exec($route?config, $request, $lookup) => router:write-response(200, $route?config)
+        util:log("info", ``[[`{$request-data?id}`] request `{$request-data?method}` `{$request-data?path}`]``),
+        try {
+            (: load router definitions :)
+            let $specs := 
+                for $api-file in $api-files
+                let $file-path := concat($base-collection, $api-file)
+                let $spec := json-doc($file-path)
+                return 
+                    if (exists($spec))
+                    then ($spec)
+                    else error($errors:OPERATION, "Failed to load JSON file from " || $file-path)
+
+            (: find all matching routes :)
+            let $matching-routes :=
+                for $spec at $pos in $specs
+                let $priority := $pos
+                return
+                    router:match-route($request-data, $spec, $priority)
+
```

Comment:
Checking every route every time costs performance. I can see an increased load on `router:match-route` and `router:create-regex`. I would thus still test for the request method first to limit the number of regex tests to perform.

Reply:
It is strictly necessary to have it checked in that order in order to satisfy the specification. Otherwise error 405 cannot be identified: a matching route but an unsupported method.
In order to gain speed, I have a suggestion:
In `router:match-route`:

- extract the static part of each route with `substring-before($pattern, "{")`
- before even creating the regular expression check if the request path begins with the static part
- moving all routes without a static part to the end of the definition

Reply:
Good suggestion. It's not very urgent as I'm not seeing major performance drops in my stress tests right now. But might be good to have this optimization.